### PR TITLE
Add Telegram Desktop

### PR DESCRIPTION
**Pull request by abbradar**
###### Things done:
- [x] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
###### More

Adds Telegram Desktop and required packages.

P.S.: In my personal rating of packages with completely crazy, "because f*ck you that's why" build sequences this would make a top, and for a long time I feel. The most funny thing is that, modulo several improvements for using more system libraries and more `INCLUDEPATH`s than expected (Because why use pkg-config? Let's just hardcode FHS paths for all packages we use and their dependencies by hand!), this is _an official building procedure_ (see [1](https://github.com/telegramdesktop/tdesktop/blob/master/doc/building-qmake.md)). _Yes_, with a statically built Qt with custom patches "because our font rendering engine uses internal Qt calls". _Yes_, with sed-patches and manual adding of `LIBS` to link to. _Yes_, with this `grep ... | sed ... | xargs make` -- because why spend time to figure out how to properly list dependencies in QMake? And even -- _yes_, with Arch AUR package as an official source for .desktop files.

Not a nice package, certainly. If you spot any profanity in the code, please excuse me -- I was forced.

### CODE REVIEW

File: pkgs/applications/networking/instant-messengers/telegram/tdesktop/default.nix [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    sed "s,/usr/bin,$out/bin,g" $tgaur/tg.protocol > $out/share/kde4/services/tg.protocol
+    for icon_size in 16 32 48 64 128 256 512; do
+      install -Dm644 "Telegram/SourceFiles/art/icon''${icon_size}.png" "$out/share/icons/hicolor/''${icon_size}x''${icon_size}/apps/telegram-desktop.png"
+    done
+
+    fixupPhase
+  '';
+
+  meta = with stdenv.lib; {
+    description = "Utility library providing GObject-based interfaces and classes for commonly used data structures";
```
Comment (7c6f434c) [inline pkgs/applications/networking/instant-messengers/telegram/tdesktop/default.nix]: I hope this line is a typo.

Comment (abbradar) [inline pkgs/applications/networking/instant-messengers/telegram/tdesktop/default.nix]: Copy-paste error. Thanks, fixed!

**Comment by 7c6f434c** [comment]
I hope `pcre` path didn't change in the `pcre8` variant? Looks good to me.

**Comment by abbradar** [comment]
Yes, I've taken extra care to ensure plain `pcre` is not affected. Travis error is due to log growing too much. I'll merge this in several days to leave a window for review.

**Comment by astsmtl** [comment]
Official Telegram client doesn't support alternative input methods (like uim). It seems like this is due to the fact they use their own version of Qt. Do you have any idea how to fix this?

**Comment by abbradar** [comment]
We can try to patch it to use proper system Qt, but that can quickly turn tricky. See https://github.com/telegramdesktop/tdesktop/issues/1815

**Comment by jagajaga** [comment]
cc @volhovM

**Comment by jgillich** [comment]
> our font rendering engine uses internal Qt calls

Waow. Because a chat app is so complicated, of course you can't do that with the regular Qt API.

**Pull request merged**